### ISSUE TITLE
bug in RDA eigenvalues

### ISSUE THREAD

**Issue by jairideout** [biocore/scikit-bio#1002]
See [here](https://github.com/biocore/scikit-bio/pull/925#issuecomment-113173055) for some discussion. @Jorge-C are you available to work on this?

**Comment by jairideout** [comment]
@mortonjt are you available to work on this for 0.5.0? We're shooting for June 7 to cut the release.

**Comment by mortonjt** [comment]
That's a pretty tight deadline - particularly with finals week approaching
...

If the deadline were couple days later, I'd be more inclined to try to
knock out this.

**Comment by jairideout** [comment]
I'm okay with pushing the release date back a couple of days. @gregcaporaso @ebolyen any objections?

**Comment by gregcaporaso** [comment]
No objections.

**Comment by jairideout** [comment]
@mortonjt we're now shooting for Friday 6/10 as the release date, does that work for you?

**Comment by mortonjt** [comment]
My final is on the 8th.  Let me see what I what I can do by this weekend.
That might be OK.

**Comment by jairideout** [comment]
Sounds good, let us know if you'll need more time

**Comment by mortonjt** [comment]
Ok, I think I figured out exactly what is going on in vegan.

In vegan, they are normalizing the eigenvalues by the number of samples

So their eigenvalues reported are actually as follows


```python
eigvals = np.sqrt(s / n-1)
```

Where `s` is the eigenvalues from the `np.linalg.svd`,  `n` is the number of samples, and `eigvals` is the reported eigenvalues.

TL;DR  I think we were are doing here is perfectly legit.  Its just a matter of personal preference.  If we want to be consistent with vegan, the change is really easy.  However, I don't see the benefit of doing so.

**Comment by antgonza** [comment]
What about adding an option, by default set to False, so the results match? Does vegan's manual state that deviation?

**Comment by mortonjt** [comment]
👎 Its off by a constant, which can be easily corrected by the user if need be.  Adding an extra flag would add unnecessary complexity to the API.

I think we should opt for the vegan approach, or use the existing approach.  We can add documentation if necessary, explaining exactly what happened.  Thoughts?

**Comment by mortonjt** [comment]
And no, vegan doc's didn't state this at all.  It required a little guesswork to validate this.

**Comment by antgonza** [comment]
OK, what about adding two methods: rda, and rda_vegan, similar to beta divs: bray_curtis, bray_curtis_faith, bray_curtis_magurran? Agree on adding documentation about this difference is important.

**Comment by jairideout** [comment]
Since it's just off by a constant, how about updating the docs for now to describe how the eigenvalues are different from vegan, and how to obtain vegan's normalized eigenvalues if that's desired? Probably worth noting that scikit-bio's eigenvalues match Legendre & Legendre.

**Comment by ebolyen** [comment]
:+1: for updating docs

**Comment by antgonza** [comment]
Only updating the documentation is one solution and should be done either way but not having both implementations means that users will have to use R vs. skbio to reproduce plots.

**Comment by jairideout** [comment]
👎, we already deviate from R/vegan API in several places. scikit-bio isn't a replacement for R/vegan, and if R behavior is desired, use R. vegan is one of many R packages implementing ecological statistics, and they all have different APIs and (sometimes) different results. If we implement functions to match vegan exactly, we'd also need to implement functions that match any number of other R packages (e.g., the Mantel test can be found in vegan, ecodist, ade4, ape, ncf, probably others).

I think where practical we should match vegan or whatever the most trusted source is. In this case, it seems more justified to match Legendre & Legendre because vegan is normalizing eigenvalues differently and it isn't documented why. If users want vegan's normalized eigenvalues, scikit-bio's docs will describe how to do that.

**Comment by wasade** [comment]
:+1: on docs. IMO, a clear note on how it differs from vegan is sufficient.

**Comment by rob-knight** [comment]
Agree with Jai. Ideally we will write a series of short papers
identifying/benchmarking such discrepancies and justifying why we use the
optimal one in skbio, but skbio should not be the variorum edition of every
method tweak in microbial ecology...

Note that the microbial ecology indices and analyses, although we typically
use them now for amplicon data, will need to be repurposed and
re-benchmarked for metabolomics, shotgun metagenomics etc. so there is a
lot of scope for exciting new work on existing methods.

On Jun 3, 2016, at 5:12 PM, Jai Ram Rideout <notifications@github.com>
wrote:

👎, we already deviate from R/vegan API in several places. scikit-bio isn't
a replacement for R/vegan, and if R behavior is desired, use R. vegan is
one of many R packages implementing ecological statistics, and they all
have different APIs and (sometimes) different results. If we implement
functions to match vegan exactly, we'd also need to implement functions
that match any number of other R packages (e.g., the Mantel test can be
found in vegan, ecodist, ade4, ape, ncf, probably others).

I think where practical we should match vegan or whatever the most trusted
source is. In this case, it seems more justified to match Legendre &
Legendre because vegan is normalizing eigenvalues differently and it isn't
documented why. If users want vegan's normalized eigenvalues, scikit-bio's
docs will describe how to do that.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/biocore/scikit-bio/issues/1002#issuecomment-223716001>,
or mute the thread
<https://github.com/notifications/unsubscribe/AB69Kc6zTOvBdg_uHGcBbz_b6IMN2YcQks5qILTIgaJpZM4FWr42>
.

**Comment by jairideout** [comment]
@mortonjt, I noticed `skbio/stats/ordination/tests/test_redundancy_analysis.py` has a few comments of the form `# These are wrong. See issue #1002`. When the docs are updated to "fix" this issue, can you also remove those comments from the code or update them to note why the expected eigenvalues are different?

**Comment by jairideout** [comment]
Fixed in #1370, thanks @mortonjt!
